### DISCUSSION TITLE
Is there a limit in how many valid redirects you an add?

### DISCUSSION THREAD

**Discussion by kristianbendren** [keycloak/keycloak#9185] [Q&A]
As keycloak is not support asterixs in valid redirects we need to add about 5000 valid redirects for our subdomains and in the future it will be many more.
Is this an issue or will it work by default?

Also is API supporting adding valid redirects?

**Comment by Sepax** [comment]
I'm wondering this too.

**Comment by stianst** [comment]
So, you're saying that for a single application you have 5000 subdomains for the same application? That doesn't sound quite right to me.

**Reply by bs-matil** [reply]
There are some dynamic use cases for this. 
A good example is Eclipse Che.
Che can maintain dynamic workspaces for users. To ingress the traffic for a workspace there are multiple ways. One is to create dynamic subdomains for each created workspace and than route the traffic DNS based (which has a lot of advantages for the operational part) 

You will then get something like `<uuid>.workspaces.myhost.com` or similar. This workspaces are basically all the same application/client. If a user wants to login in his workspace he will be redirected to his workspace, which only works if he gets is own client  or a shared client with state astrix. 

I don't have any opinion if this is a good or bad thing. But there are applications build like this.

**Reply by stianst** [reply]
Hmm.. makes sense. Problem is of course that there are also many more situations where subdomains refer to different applications, and also many cases where different parties maintain the subdomains.

The proper solution to this, which is what we have proposed several times is to leverage client policies to make it possible to set policies on what redirect-uris are permitted. Where the options would be something like strict (default), wildcard-in-subdomain, wildcard-in-path, etc.. That will allow an admin to configure strict for all applications, except for only the few that need to support wildcard in subdomain.

Bear in mind one thing here - best practices from OAuth 2 and OAuth 2.1 requires strict redirect-uri matching.

**Reply by bs-matil** [reply]
I fully agree this should not be possible by default and if available used with caution. Its nothing which I would expect to be used widely.

**Reply by kristianbendren** [reply]
In our case each customer gets a subdomain for their site. It's solved by routing in our application so it's not an real subdomain in DNS records.
So for each customer site they need to be able to login into their account and have the redirect back.

I would also say that allowing wildcard subdomain in redirect uri would solve this but what i understood it wont happen.
So our solution is to then add all subdomains for each customer.

**Reply by stianst** [reply]
@krillehbg Not sure why you think supporting wildcard in subdomains won't happen. All we're saying is that we are moving in the direction of not allowing wildcards in general, and would like to leverage client policies to allow admins to define what clients should be permitted to use wildcards where.

**Reply by kristianbendren** [reply]
@stianst im refering to this pull request that was closed

https://github.com/keycloak/keycloak/pull/8187
I would be more than glad if this will be supported as it doesn't feel good to have all those redirects imported.

**Reply by stianst** [reply]
@krillehbg that PR was closed due to the approach, not as a general thing against wildcards in subdomains. See here for a follow up https://github.com/keycloak/keycloak/pull/9281

**Reply by bs-matil** [reply]
I think @stianst linked the wrong document 
The correct one is #9278 I guess.

**Comment by stianst** [comment]
I've started a discussion around the redirect-uris and what we should do here in the future:
https://github.com/keycloak/keycloak/discussions/9278
